dht_bootstrap.go

```diff
@@ -163,22 +109,68 @@ func (dht *IpfsDHT) runBootstrap(ctx context.Context, cfg BootstrapConfig) error
 		return err
 	}
 
-	// Do all but one of the bootstrap queries as random walks.
-	for i := 0; i < cfg.Queries; i++ {
-		err := doQuery(i, "random ID", dht.randomWalk)
-		if err != nil {
-			return err
+	buckets := dht.routingTable.GetAllBuckets()
+	var wg sync.WaitGroup
+	errChan := make(chan error)
+
+	for bucketID, bucket := range buckets {
```

Comment:
We should stagger these lookups, instead of launching them all at once. This can explode at the dialer.

Reply:
@raulk  Please can you explain this in context of https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r334237157 ?

Reply:
I might be missing something, because in my head this logic here is not related to #383. I'll explain what I mean anyway. If we have 10 buckets to refresh, this logic performs 10 "find peer" queries, which multiplies into as many as 200 simultaneous dials (worst case scenario), as permitted by the dial queue. I don't think that will scale well.

Reply:
Pinging @Stebalien ^^

Reply:
@raulk Oh, I was referring to a [specific comment](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r334237157) in 383. Since you mention that the throttling logic to restrict simultaneous dials already exists in the dialer, I was wondering why the dialer would blow up here. Or did you have something else in mind for the comment on 383 ?

Reply:
I wouldn't expect 10 concurrent requests to be an issue. The dialer overload issues came from the fact that the gateways are making many concurrent DHT requests _per user request_. That means 1000s of DHT requests.

We _could_ stagger them when running them in the background but:

1. I'm not sure if it's really worth it.
2. We definitely want to fire them all off as fast as possible on start.
3. We only need to bootstrap when we're _not_ under heavy DHT load. When we're under load, the buckets will all be "fresh" and this code will be a no-op.

Reply:
10 concurrent requests x 20 max. pending dials in dial queue = potentially 200 outstanding dials, which is larger than the default FD limit we ship with (160 dials). In the worst case scenario, the DHT could max out the dialer.